src/v/cluster/rm_stm.cc

```diff
@@ -423,11 +423,45 @@ ss::future<tx_errc> rm_stm::commit_tx(
     co_return tx_errc::none;
 }
 
+rm_stm::abort_origin rm_stm::get_abort_origin(
+  const model::producer_identity& pid, const model::tx_seq& tx_seq) const {
+    auto expected_it = _mem_state.expected.find(pid);
+    if (expected_it != _mem_state.expected.end()) {
+        if (tx_seq < expected_it->second) {
+            return abort_origin::past;
+        } else if (expected_it->second < tx_seq) {
+            return abort_origin::future;
+        }
+    }
+
+    auto preparing_it = _mem_state.preparing.find(pid);
+    if (preparing_it != _mem_state.preparing.end()) {
+        if (tx_seq < preparing_it->second.tx_seq) {
+            return abort_origin::past;
+        } else if (preparing_it->second.tx_seq < tx_seq) {
+            return abort_origin::future;
+        }
+    }
+
+    auto prepared_it = _log_state.prepared.find(pid);
+    if (prepared_it != _log_state.prepared.end()) {
```

Comment:
nit: there is a syntax you might find convenient because you don't have a (1) choose a non-conflicting iterator name and (2) doesn't populate the scope with extra names.

```c++
if (auto it = _log_state.prepared.find(pid); it != _log_state.prepared.end()) {
  if (tx_seq < it->second.tx_seq) {
...
}
```

Reply:
thanks, I've seen it. the problem with this syntax is than `len("if (auto it = _log_state.prepared.find(pid); it != _log_state.prepared.end()) {")` is already 80 chars, when we add the indent we'll need to wrap and I find it harder to read if-conditions when they are not on the same line so I prefer the former form